Add SendGrid webhook event models and signature verification

thoven87 opened the pull request:
Introduces comprehensive models for SendGrid webhook event types, including delivery, engagement, and account status change events. Adds ECDSA signature verification for webhook security, updates Package.swift to include swift-crypto, and provides extensive documentation and tests for event parsing and verification.

Closes #28

thoven87 committed:
Add SendGrid webhook event models and signature verification

Introduces comprehensive models for SendGrid webhook event types, including delivery, engagement, and account status change events. Adds ECDSA signature verification for webhook security, updates Package.swift to include swift-crypto, and provides extensive documentation and tests for event parsing and verification.

fpseverino commented:
So this basically only adds structs that a Vapor/Hummingbird server should decode and methods to verify their signature, right?

thoven87 commented:
> So this basically only adds structs that a Vapor/Hummingbird server should decode and methods to verify their signature, right?

I just noticed that I am missing the received case. I will fix this shortly along with the CRUD operations https://www.twilio.com/docs/sendgrid/api-reference/webhooks/create-an-event-webhook

thoven87 committed:
Add Event and Parse Webhook API support

Introduces methods for managing SendGrid Event Webhooks (create, test, retrieve, update, delete, signature verification, public key retrieval) and Parse Webhook settings/statistics in SendGridClient. Also updates SendGridError to include url and id fields for improved error context, and fixes the base API URL to support new endpoints.

thoven87 committed:
Add support for received webhook events and settings

Introduces SendGridReceivedEvent to handle inbound email 'received' events, updates SendGridWebhookEvent and related enums for the new event type, and adds models for event webhook input, settings, signature verification, and parse webhook statistics. Also adds sgContentType to engagement events and improves documentation for uniqueArgs.

thoven87 committed:
Add tests for SendGrid received event decoding

Added unit tests to verify decoding of SendGrid 'received' webhook events, including support for both Unix and ISO 8601 timestamps, additional fields, and v3_payload_details. Also updated existing test data to use generic test values for consistency.

thoven87 committed:
Refactor method signatures for readability

Updated several public method signatures in SendGridClient to use multiline formatting for parameters, improving readability and consistency. Also adjusted body encoding and decoding calls to multiline style.

thoven87 committed:
Update expected sgEventId in delivered event test

Changed the expected sgEventId value in the delivered event decoding test to 'test_delivered_event_id' for consistency or improved test clarity.

thoven87 commented:
> So this basically only adds structs that a Vapor/Hummingbird server should decode and methods to verify their signature, right?

@fpseverino all the failing tests are related to the email validation API

fpseverino commented on README.md:
````diff
@@ -160,3 +160,194 @@ do {
+
+```swift
+import SendGridKit
+import NIO
````
This should be Vapor, right? The example uses Vapor APIs it seems

thoven87 replied:
Yes, I will look at the Vapor doc to update this example.

fpseverino commented on README.md:
````diff
@@ -160,3 +160,194 @@ do {
 }
 ```
+
+### Webhook Event Processing
````
Could you also add this excellent documentation as an article in the DocC catalog?

thoven87 replied:
I'll do so shortly.

thoven87 replied:
> Could you also add this excellent documentation as an article in the DocC catalog?

This is now completed.

fpseverino commented on Sources/SendGridKit/Models/WebhookEvent.swift:
```diff
@@ -0,0 +1,889 @@
+        public let id: Int
+    }
+
+    public struct Newsletter: Codable, Sendable {
+        public let newsletterUserListId: String
+        public let newsletterId: String
+        public let newsletterSendId: String
+
+        enum CodingKeys: String, CodingKey {
+            case newsletterUserListId = "newsletter_user_list_id"
+            case newsletterId = "newsletter_id"
+            case newsletterSendId = "newsletter_send_id"
+        }
+    }
```
A minor stylistic issue: in this package, when the term ID appears together with other words in a property, we always capitalise both letters.
For example:
```suggestion
    public struct Newsletter: Codable, Sendable {
        public let newsletterUserListID: String
        public let newsletterID: String
        public let newsletterSendID: String

        enum CodingKeys: String, CodingKey {
            case newsletterUserListID = "newsletter_user_list_id"
            case newsletterID = "newsletter_id"
            case newsletterSendID = "newsletter_send_id"
        }
    }
```

fpseverino commented on Sources/SendGridKit/Models/WebhookEvent.swift:
```diff
@@ -0,0 +1,889 @@
+            throw DecodingError.dataCorrupted(
+                DecodingError.Context(
+                    codingPath: decoder.codingPath,
+                    debugDescription: "Could not determine event type"
```
```suggestion
                    debugDescription: "Could not determine SendGrid webhook event type"
```

fpseverino commented on Sources/SendGridKit/Models/WebhookEvent.swift:
```diff
@@ -0,0 +1,889 @@
+    /// Type of bounce or status change.
+    public let type: StatusType?
+    /// Custom arguments passed with the message.
+    /// SendGrid documentation states that unique_args should only contain string values.
```
```suggestion
    /// SendGrid documentation states that `unique_args` should only contain string values.
```

fpseverino commented on Sources/SendGridKit/Models/WebhookEvent.swift:
```diff
@@ -0,0 +1,889 @@
+/// Delivery events: bounce, delivered, deferred, dropped, processed, blocked
+public struct SendGridDeliveryEvent: Codable, Sendable {
+    /// ID of the unsubscribe group that includes the recipient's email address.
+    public let asmGroupId: Int?
```
Same thing here with all the properties that end with "Id"

fpseverino commented on Sources/SendGridKit/Models/WebhookEvent.swift:
We need to split this file into smaller ones, you should create a subfolder (inside the "Models" one) called "Webhooks" or something like that

thoven87 replied:
> We need to split this file into smaller ones, you should create a subfolder (inside the "Models" one) called "Webhooks" or something like that

I don't know how much smaller you want. I split the file into two categories, Input and Response

fpseverino commented on Sources/SendGridKit/Webhook+SignatureVerification.swift:
```diff
@@ -0,0 +1,122 @@
+        if !isValid {
+            throw SendGridWebhookSignatureError.noMatchingSignatureFound
+        }
+    }
```
```suggestion
        guard ecdsaPublicKey.isValidSignature(ecdsaSignature, for: SHA256.hash(data: signedData)) else {
            throw SendGridWebhookSignatureError.noMatchingSignatureFound
        }
    }
```

fpseverino commented on Sources/SendGridKit/Webhook+SignatureVerification.swift:
```diff
@@ -0,0 +1,122 @@
+extension Data {
+    /// Helper to convert Data to hex string
+    var hexString: String {
+        return self.map { String(format: "%02x", $0) }.joined()
```
```suggestion
        self.map { "0\(String($0, radix: 16))".suffix(2) }.joined()
```
`printf` formats are not a good idea in Swift

fpseverino commented on Sources/SendGridKit/SendGridClient.swift:
```diff
@@ -56,4 +62,373 @@ public struct SendGridClient: Sendable {
+    /// - Parameters:
+    ///   - input: The request body
+    ///   - onbehalfOf: The on-behalf-of header allows you to make API calls from a parent account on behalf of the parent's Subusers or customer accounts.
+    public func createEventWebhook(_ input: EventWebhookInput, onbehalfOf: String? = nil) async throws -> WebhookSettingsResponse {
```
I think we should make a separate client, just like for the email validation API

thoven87 replied:
Fair enough! I will do that.

fpseverino reviewed (changes requested):
A very well designed API, great work!
You tested it with real SendGrid API keys, right?

thoven87 commented:
> A very well designed API, great work!
> 
> You tested it with real SendGrid API keys, right?

Yes!

thoven87 committed:
Update WebhookEvent.swift

Co-authored-by: Francesco Paolo Severino <96546612+fpseverino@users.noreply.github.com>

thoven87 committed:
Update WebhookEvent.swift

Co-authored-by: Francesco Paolo Severino <96546612+fpseverino@users.noreply.github.com>

thoven87 committed:
Update WebhookEvent.swift

Co-authored-by: Francesco Paolo Severino <96546612+fpseverino@users.noreply.github.com>

thoven87 committed:
Update Webhook+SignatureVerification.swift

Co-authored-by: Francesco Paolo Severino <96546612+fpseverino@users.noreply.github.com>

thoven87 committed:
Update Webhook+SignatureVerification.swift

Co-authored-by: Francesco Paolo Severino <96546612+fpseverino@users.noreply.github.com>

thoven87 committed:
Fix newsletter property names in webhook event tests

thoven87 committed:
Refactor webhook API to SendGridWebhookClient

Moved all Event Webhook-related API methods from SendGridClient to a new SendGridWebhookClient for better separation of concerns. Added public initializers to model structs for easier instantiation. Updated tests to use the new SendGridWebhookClient.

thoven87 committed:
Rename SendGrid event model properties to use ID suffix

Standardizes property names in SendGrid event models by renaming fields such as asmGroupId to asmGroupID, sgEventId to sgEventID, and similar fields to consistently use 'ID' suffix capitalization. Updates all related CodingKeys, initializers, encoders/decoders, and test cases to match the new naming convention for improved clarity and Swift API guidelines compliance.

thoven87 committed:
Move webhook input models to separate file

Extracted EventWebhookInput and SendGridTestWebhookInput structs from WebhookEvent.swift into a new file WebhookEvent+Input.swift for better code organization and separation of concerns.

thoven87 committed:
Add comprehensive SendGrid webhook documentation

Introduces detailed .docc documentation for SendGridKit webhook types, including EventWebhookInput, SendGridTestWebhookInput, SendGridWebhookClient, SendGridWebhookEvent, and signature verification. Updates the main documentation index to reference new webhook types and corrects a minor typo in SendGridWebhookClient's initializer comment. Also updates the README to use Vapor for webhook examples and clarifies body handling.
